sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala

```diff
@@ -0,0 +1,166 @@
+ *      be a lambda function defined in an outer scope, or a attribute in produced by the plan's
+ *      child. If names are duplicate, the name defined in the most inner scope is used.
+ */
+case class ResolveLambdaVariables(conf: SQLConf) extends Rule[LogicalPlan] {
+
+  type LambdaVariableMap = Map[String, NamedExpression]
+
+  private val canonicalizer = {
+    if (!conf.caseSensitiveAnalysis) {
+      s: String => s.toLowerCase
+    } else {
+      s: String => s
+    }
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    plan.resolveOperators {
+      case q: LogicalPlan =>
+        q.mapExpressions(resolve(_, Map.empty))
+    }
+  }
+
+  /**
+   * Create a bound lambda function by binding the arguments of a lambda function to the given
+   * partial arguments (dataType and nullability only). If the expression happens to be an already
+   * bound lambda function then we assume it has been bound to the correct arguments and do
+   * nothing. This function will produce a lambda function with hidden arguments when it is passed
+   * an arbitrary expression.
+   */
+  private def createLambda(
+      e: Expression,
+      partialArguments: Seq[(DataType, Boolean)]): LambdaFunction = e match {
+    case f: LambdaFunction if f.bound => f
+
+    case LambdaFunction(function, names, _) =>
+      if (names.size != partialArguments.size) {
+        e.failAnalysis(
+          s"The number of lambda function arguments '${names.size}' does not " +
+            "match the number of arguments expected by the higher order function " +
+            s"'${partialArguments.size}'.")
+      }
+
+      if (names.map(a => canonicalizer(a.name)).distinct.size < names.size) {
+        e.failAnalysis(
+          "Lambda function arguments should not have names that are semantically the same.")
+      }
+
+      val arguments = partialArguments.zip(names).map {
+        case ((dataType, nullable), ne) =>
+          NamedLambdaVariable(ne.name, dataType, nullable)
+      }
+      LambdaFunction(function, arguments)
+
+    case _ =>
+      // This expression does not consume any of the lambda's arguments (it is independent). We do
+      // create a lambda function with default parameters because this is expected by the higher
+      // order function. Note that we hide the lambda variables produced by this function in order
+      // to prevent accidental naming collisions.
+      val arguments = partialArguments.zipWithIndex.map {
+        case ((dataType, nullable), i) =>
+          NamedLambdaVariable(s"col$i", dataType, nullable)
+      }
+      LambdaFunction(e, arguments, hidden = true)
+  }
+
+  /**
+   * Resolve lambda variables in the expression subtree, using the passed lambda variable registry.
+   */
+  private def resolve(e: Expression, parentLambdaMap: LambdaVariableMap): Expression = e match {
+    case _ if e.resolved => e
+
+    case h: HigherOrderFunction if h.inputResolved =>
```

Comment:
can add some basic type check here? Then we can fail fast if the `ArrayTransform#input` is not array type, and we don't need the hacky workaround in `ArrayTransform#bind`

Reply:
Let me think about it later.